### PR TITLE
Document ait; first stab at concurrency_aware_ait.

### PR DESCRIPTION
**Pull request by johanneskloos**
ait, ait_baset and related classes all have documentation now.

The full documentation of concurrency_aware_ait needs input from
someone who really understands the algorithm.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/analyses/ai.h [hunk 1]
```diff
@@ -425,18 +470,23 @@ class ait:public ai_baset
...
       static_cast<const domainT &>(src), from, to);
   }
 
+  /// Make a copy of \p s.
   std::unique_ptr<statet> make_temporary_state(const statet &s) override
   {
     return util_make_unique<domainT>(static_cast<const domainT &>(s));
   }
 
+  /// Internal: implementation of the fixpedoint function using
```
Comment (peterschrammel) [inline src/analyses/ai.h]: fixed point

File: src/analyses/ai.h [hunk 2]
```diff
@@ -354,6 +354,41 @@ class ai_baset
   virtual std::unique_ptr<statet> make_temporary_state(const statet &s)=0;
 };
 
+/// Base class for static analyses. Any actual static analysis
+/// must (a) inherit from this class and (b) provide a domain class as a type argument,
+/// which must, in turn, inherit from \ref ai_domain_baset.
+///
+/// From a user's perspective, this class provides three main classes of functions:
+///
+/// 1. Running a static analysis, via \ref ai_baset#operator()(const irep_idt&, const goto_programt&, const namespacet&),
```
Comment (chrisr-diffblue) [inline src/analyses/ai.h]: Are lint/clang-format happy with the line length here?

File: src/analyses/ai.h [hunk 3]
```diff
@@ -354,6 +354,41 @@ class ai_baset
...
+/// and use `operator[]` afterwards to retrieve the results. The fixed point algorithm
+/// used is a standard worklist algorithm; the current implementation is
+/// flow- and path-sensitive, but not context-sensitive.
+///
+/// From an analysis developer's perspective, an analysis is implemented by
+/// inheriting from this class (or, if a concurrency-sensitive analysis is
+/// required, from \ref concurrency_aware_ait), providing a class implementing
+/// the abstract domain as the type for the \p domainT parameter. Most of the
+/// actual analysis functions (in particular, the minimal element, the lattice
+/// join and the state transformer) are supplied using \p domainT.
```
Comment (chrisr-diffblue) [inline src/analyses/ai.h]: "join and" -> "join, and"

**Review by chrisr-diffblue — APPROVED** [review]

File: src/analyses/ai.h [hunk 4]
```diff
@@ -397,24 +435,31 @@ class ait:public ai_baset
...
 
 protected:
+  /// Map from locations to domain elements, for the results of a static analysis.
   typedef std::
     unordered_map<locationt, domainT, const_target_hash, pointee_address_equalt>
       state_mapt;
   state_mapt state_map;
 
+  /// Look up the analysis state for a given location, instantiating a new state if required.
+  /// Used internally by the analysis.
```
Comment (owen-mc-diffblue) [inline src/analyses/ai.h]: Remove the existing documentation below, which is superseded by your new documentation

File: src/analyses/ai.h [hunk 5]
```diff
@@ -397,24 +435,31 @@ class ait:public ai_baset
...
   // this one creates states, if need be
   virtual statet &get_state(locationt l) override
   {
     return state_map[l]; // calls default constructor
   }
 
+  /// Look up the analysis state for a given location, throwing an exception if no state is
+  /// known.
+  /// Used internally by the analysis.
```
Comment (owen-mc-diffblue) [inline src/analyses/ai.h:466]: Remove the existing documentation, which is superseded by your new documentation

**Review by owen-mc-diffblue — APPROVED** [review]
Very clear

File: src/analyses/ai.h (hunk 2 above)
Comment (johanneskloos) [reply src/analyses/ai.h]: Reformatted, there's no more long lines left.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0f7f11d80b6dff9449bfea4a3e81aebf1528c323).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99821693

File: src/analyses/ai.h [hunk 6]
```diff
@@ -354,7 +354,48 @@ class ai_baset
   virtual std::unique_ptr<statet> make_temporary_state(const statet &s)=0;
 };
 
-// domainT is expected to be derived from ai_domain_baseT
+/// Base class for static analyses. Any actual static analysis
```
Comment (tautschnig) [inline src/analyses/ai.h]: Can we use "abstract interpretation" or "data-flow analysis"?

File: src/analyses/ai.h [hunk 7]
```diff
@@ -354,7 +354,48 @@ class ai_baset
...
+/// must (a) inherit from this class and (b) provide a domain class as a
+/// type argument, which must, in turn, inherit from \ref ai_domain_baset.
+///
+/// From a user's perspective, this class provides three main classes of
```
Comment (tautschnig) [inline src/analyses/ai.h]: "three groups of functions" or "three main kinds of functions"

File: src/analyses/ai.h [hunk 8]
```diff
@@ -354,7 +354,48 @@ class ai_baset
...
+/// methods:
+/// - \ref ait#initialize(const irep_idt&, const goto_programt&),
+///   \ref ait#initialize(const irep_idt&,
+///   const goto_functionst::goto_functiont&) and
+///   \ref ait#initialize(const goto_functionst&), for pre-analysis
+///   initialization
+/// - \ref ait#finalize(), for post-analysis cleanup.
+///
+// For the <!-- --> trick above, see
+// https://stackoverflow.com/questions/46744573/
```
Comment (tautschnig) [inline src/analyses/ai.h]: Maybe that should instead go in the commit message? (It's very nice indeed, we've got a few way-too-long-lines.)

File: src/analyses/ai.h [hunk 9]
```diff
@@ -457,6 +513,12 @@ class ait:public ai_baset
   }
 };
 
+/// Base class for concurrency-aware static analyses. See
+/// \ref ait for details.
+/// The only difference is that after the sequential fixed point construction,
+/// as done by \ref ait, another step is added to account for
+/// concurrently-executed instructions.
+/// TODO: Somebody who truly understands this algorithm should document it!
```
Comment (tautschnig) [inline src/analyses/ai.h]: It's basically "Analysis of Multithreaded Programs" by Martin Rinard, SAS 2001. And that is basically making the analysis flow-insensitive by feeding results of a sequential execution back into the entry point, thereby accounting for any values computed by different threads.

**Review by tautschnig — APPROVED** [review]

File: src/analyses/ai.h (hunk 9 above)
Comment (johanneskloos) [reply src/analyses/ai.h]: I took the liberty of copying your description and linking to the paper.

**Comment by owen-mc-diffblue** [comment]
@johanneskloos I think you may have force-pushed the wrong branch - none of the updates have been made in what I can see

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 74b9e4590fe425b35287df8eec4bb718a7eb0f61).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99921983
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f6b24fbf78f8f41d09cdc507feddaca34fa342d0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99925470

**Pull request merged**